git/revlist.go

```diff
@@ -117,3 +120,207 @@ func revListCallback(c *Client, opt RevListOptions, commits []CommitID, excludeL
 	}
 	return nil
 }
+
+// Internal pre-computed record of which symbolic references
+//  are pointing to a particular ref name (without the refs/ prefix)
+var symRefsForRefName map[string]string
+
+// Internal pre-computed record of which ref names are pointing to
+//  to a particular commit. Values are in the form of "HEAD -> master, branch2,
+//  tag: tag1"
+var refNamesForCommit map[string]string
+
+// We have been asked to pre-compute the ref names for important
+//  commits as a way to improve performance.
+func buildRefNames(c *Client) error {
+	symRefsForRefName = make(map[string]string)
+	refNamesForCommit = make(map[string]string)
+
+	// TODO support more symbolic refs than just HEAD
+	headrefspec, err := SymbolicRefGet(c, SymbolicRefOptions{Short: true}, "HEAD")
+	if err != nil {
+		return err
+	}
+	symRefsForRefName[string(headrefspec)] = "HEAD"
+
+	branches, err := c.GetBranches()
+	if err != nil {
+		return err
+	}
+
+	for _, branch := range branches {
+		branchName := string(branch)
+
+		commitId, err := RefSpec(branchName).Value(c)
+		if err != nil {
+			//return err There are problems dealing with hierarchical branches
+			continue
+		}
+
+		branchName = branchName[len("refs/heads/"):]
+
+		if entry, ok := symRefsForRefName[branchName]; ok {
+			branchName = entry + " -> " + branchName
+		}
+
+		if entry, ok := refNamesForCommit[commitId]; ok {
+			refNamesForCommit[commitId] = entry + ", " + branchName
+		} else {
+			refNamesForCommit[commitId] = branchName
+		}
+	}
+
+	tagNames, err := TagList(c, TagOptions{}, []string{})
+	if err != nil {
+		return err
+	}
+
+	for _, tagName := range tagNames {
+		tag := RefSpec("refs/tags/" + tagName)
+		commitId, err := tag.Value(c)
+		if err != nil {
+			//return err There are problems dealing with hierarchical tags
+			continue
+		}
+
+		if entry, ok := refNamesForCommit[commitId]; ok {
+			refNamesForCommit[commitId] = entry + ", tag: " + tagName
+		} else {
+			refNamesForCommit[commitId] = "tag: " + tagName
+		}
+	}
+
+	return nil
+}
+
+// Since libgit is somewhat out of our control and we can't implement
+// a fmt.Stringer interface there, we use this helper.
+
+// Print commit using "medium" format
+func printCommitMedium(c *Client, cmt CommitID, format string) error {
```

Comment:
Are these required? Can they be replaced with pre-defined formatting strings for printCommitFormat?

The only hiccup I'd see is merge parents, but (if git doesn't already have one) we can temporarily add a "print all merge commit IDs" verb until the looping stuff is there.

Reply:
Yes, I'm wondering if these can be replaced by formats. Perhaps it could all be put into a Commit type as a Format() method.